### PR TITLE
fix: relax type on object properties

### LINKED ISSUE #170: `Type 'undefined' is not assignable to type 'Schema'.ts(2345)` error with allOf+if/then+different properties
Hi, it seems to me that the typing for Schemas is too strict in a particular case. Please let me know if this is a bug in my code or yours:

## Repro
There are a lot of things going on in my schema but here is a reduced example. If `propertyA` is 1, then `propertyB` must be null. If `propertyA` is 2, then `propertyC` must be null. There are other properties present and validated at a higher level, but I'm leaving them out for this minimal example.

_side note_: I know this particular rule set can be converted to use `oneOf`, but there are other rules in the real schema that make `allOf` more straightforward.
```json
{
  "type": "object",
  "allOf": [
    {
      "if": {
        "properties": {
          "propertyA": {
            "const": 1
          }
        }
      },
      "then": {
        "properties": {
          "propertyB": {
            "const": null
          }
        }
      }
    },
    {
      "if": {
        "properties": {
          "propertyA": {
            "const": 2
          }
        }
      },
      "then": {
        "properties": {
          "propertyC": {
            "const": null
          }
        }
      }
    }
  ]
}
```

From there, I'm just importing it and using `validator`:
```ts
import mySchema from '@/schemas/mySchema.json'
import { validator } from '@exodus/schemasafe';
const myValidator = validator(mySchema) // Typescript error here
```

here's the full typescript error:
```
Argument of type '{ type: string; allOf: ({ if: { properties: { propertyA: { const: number; }; }; }; then: { properties: { propertyB: { const: null; }; propertyC?: undefined; }; }; } | { if: { properties: { propertyA: { const: number; }; }; }; then: { ...; }; })[]; }' is not assignable to parameter of type 'Schema'.
  Type '{ type: string; allOf: ({ if: { properties: { propertyA: { const: number; }; }; }; then: { properties: { propertyB: { const: null; }; propertyC?: undefined; }; }; } | { if: { properties: { propertyA: { const: number; }; }; }; then: { ...; }; })[]; }' is not assignable to type '{ $schema?: string | undefined; $vocabulary?: string | undefined; id?: string | undefined; $id?: string | undefined; $anchor?: string | undefined; $ref?: string | undefined; definitions?: { ...; } | undefined; ... 54 more ...; discriminator?: { ...; } | undefined; }'.
    Types of property 'allOf' are incompatible.
      Type '({ if: { properties: { propertyA: { const: number; }; }; }; then: { properties: { propertyB: { const: null; }; propertyC?: undefined; }; }; } | { if: { properties: { propertyA: { const: number; }; }; }; then: { properties: { ...; }; }; })[]' is not assignable to type 'Schema[]'.
        Type '{ if: { properties: { propertyA: { const: number; }; }; }; then: { properties: { propertyB: { const: null; }; propertyC?: undefined; }; }; } | { if: { properties: { propertyA: { const: number; }; }; }; then: { properties: { ...; }; }; }' is not assignable to type 'Schema'.
          Type '{ if: { properties: { propertyA: { const: number; }; }; }; then: { properties: { propertyB: { const: null; }; propertyC?: undefined; }; }; }' is not assignable to type 'Schema'.
            Type '{ if: { properties: { propertyA: { const: number; }; }; }; then: { properties: { propertyB: { const: null; }; propertyC?: undefined; }; }; }' is not assignable to type '{ $schema?: string | undefined; $vocabulary?: string | undefined; id?: string | undefined; $id?: string | undefined; $anchor?: string | undefined; $ref?: string | undefined; definitions?: { ...; } | undefined; ... 54 more ...; discriminator?: { ...; } | undefined; }'.
              Types of property 'then' are incompatible.
                Type '{ properties: { propertyB: { const: null; }; propertyC?: undefined; }; }' is not assignable to type 'Schema | undefined'.
                  Type '{ properties: { propertyB: { const: null; }; propertyC?: undefined; }; }' is not assignable to type '{ $schema?: string | undefined; $vocabulary?: string | undefined; id?: string | undefined; $id?: string | undefined; $anchor?: string | undefined; $ref?: string | undefined; definitions?: { ...; } | undefined; ... 54 more ...; discriminator?: { ...; } | undefined; }'.
                    Types of property 'properties' are incompatible.
                      Type '{ propertyB: { const: null; }; propertyC?: undefined; }' is not assignable to type '{ [id: string]: Schema; }'.
                        Property '"propertyC"' is incompatible with index signature.
                          Type 'undefined' is not assignable to type 'Schema'.ts(2345)
```

It seems like `schemasafe` expects the same properties to be named in each sub-schema, but I don't see why that would be the case.

### PR DESCRIPTION
**Pull request by sparten11740**
This PR relaxes the type on object properties to accommodate for conflicting type inference as reported in #170

A simplified version of the issue we are facing is the following:
```ts
const y = [
  { "X": true },
  { "Y": false }
]
````

Is inferred as 
```ts
({
    X: boolean;
    Y?: undefined;
} | {
    Y: boolean;
    X?: undefined;
})[]
```

`properties` amongst other is typed as `{ [id: string]: Schema }` . This index signature states that if we have a property, it has to be of type `Schema` and cannot be `undefined`. This conflicts with the type that is inferred from e.g. an array provided to `allOf`, where one element references a different property than the other. Similar to the example above one property in one branch of the union type will then be inferred as `propA?: undefined`. 

A possible solution is to allow keys with `undefined` values in the index signature using either `Partial<{ [id: string]: Schema }>` or `<{ [id: string]: Schema | undefined }>`. The caveat of is that obviously a key with an `undefined` value does not make sense in `properties`. Worth paying that price?

This PR does that and adds a test case for the bug from #170

Closes #170

### CODE REVIEW

File: .github/workflows/nodejs.yml [hunk 1]
```diff
@@ -23,3 +23,16 @@ jobs:
       - run: yarn lint
       - run: git submodule update --init --recursive
       - run: yarn run test
+  types:
```
Comment (mvayngrib) [inline .github/workflows/nodejs.yml:26]: nice touch, adding these!

**Review by mvayngrib — APPROVED** [review]
utACK

**Review by ChALkeR — APPROVED** [review]
tACK

This is very nice, thanks!

File: package.json [hunk 2]
```diff
@@ -36,7 +36,8 @@
     "test": "npm run test:raw | tap-spec",
     "test:raw": "npm run test:normal && npm run test:module",
     "test:module": "tape -r ./test/tools/test-module.js test/*.js test/regressions/*.js",
-    "test:normal": "tape test/*.js test/regressions/*.js"
+    "test:normal": "tape test/*.js test/regressions/*.js",
+    "test:types": "cd test/types && yarn test"
```
[the remark this answers is not in the document]
Comment (sparten11740) [reply package.json]: simplified script in https://github.com/ExodusMovement/schemasafe/pull/180/commits/f1dd3d8f51fb7839213823ee426c8ef5cc843326

**Review by ChALkeR — APPROVED** [review]

**Review by mvayngrib — APPROVED** [review]
utACK

**Pull request merged**